src/coprocessor/trie/mod.rs

```diff
@@ -0,0 +1,734 @@
+        Self { preimage_path }
+    }
+
+    /// Verify a `LookupProof`. Note that this verification is exactly what must be proved in the circuit.
+    pub fn verify(&self, root: F, key: F, value: F, hash_cache: &PoseidonCache<F>) -> bool {
+        let path = Trie::<F, ARITY, HEIGHT>::path(key);
+
+        let mut next = root;
+        for (k, preimage) in path.iter().zip(&self.preimage_path) {
+            let computed_hash = Trie::<F, ARITY, HEIGHT>::compute_hash(hash_cache, *preimage);
+
+            if next != computed_hash {
+                return false;
+            }
+            next = preimage[*k];
+        }
+        next == value
+    }
+}
+
+pub struct InsertProof<F: LurkField, const ARITY: usize, const HEIGHT: usize> {
+    old_proof: LookupProof<F, ARITY, HEIGHT>,
+    new_proof: LookupProof<F, ARITY, HEIGHT>,
+}
+
+impl<F: LurkField, const ARITY: usize, const HEIGHT: usize> InsertProof<F, ARITY, HEIGHT> {
+    fn new(
+        old_proof: LookupProof<F, ARITY, HEIGHT>,
+        new_proof: LookupProof<F, ARITY, HEIGHT>,
+    ) -> Self {
+        Self {
+            old_proof,
+            new_proof,
+        }
+    }
+
+    /// Verify an `InsertProof`. Note that this verification is exactly what must be proved in the circuit.
+    pub fn verify(
+        &self,
+        old_root: F,
+        new_root: F,
+        key: F,
+        old_value: Option<F>,
+        new_value: F,
+        hash_cache: &PoseidonCache<F>,
+    ) -> bool {
+        let old_verified = self.old_proof.verify(
+            old_root,
+            key,
+            old_value.unwrap_or_else(|| F::zero()),
+            hash_cache,
+        );
+
+        let new_verified = self.new_proof.verify(new_root, key, new_value, hash_cache);
+
+        let paths_differ_by_at_most_one = self
+            .old_proof
+            .preimage_path
+            .iter()
+            .zip(&self.new_proof.preimage_path)
+            .all(|(a, b)| {
+                let differing_postition_count = a
+                    .iter()
+                    .zip(b)
+                    .map(|(x, y)| x != y)
+                    .fold(0, |acc, are_diff| acc + are_diff as usize);
+                differing_postition_count == 1 || differing_postition_count == 0
+            });
+
+        old_verified && new_verified && paths_differ_by_at_most_one
+    }
+}
+
+impl<'a, F: LurkField, const ARITY: usize, const HEIGHT: usize> Trie<'a, F, ARITY, HEIGHT> {
+    fn compute_hash(hash_cache: &PoseidonCache<F>, preimage: [F; ARITY]) -> F {
+        macro_rules! hash {
+            ($hash_name:ident, $n:expr) => {{
+                let mut buffer = [F::zero(); $n];
+                buffer.copy_from_slice(&preimage);
+
+                hash_cache.$hash_name(&buffer)
+            }};
+        }
+        match ARITY {
+            3 => hash!(hash3, 3),
+            4 => hash!(hash4, 4),
+            6 => hash!(hash6, 6),
+            8 => hash!(hash8, 8),
+            _ => unimplemented!(),
```

Comment:
It would be great to front-load this panic in `new`and `new_with_root`.

Reply:
Moved this helper to `HashArity` and added a check in (newly added) `new_aux`.